contracts/test/integration/L2PlusXNet.t.sol

```diff
@@ -0,0 +1,599 @@
+    bytes32 private new_topdown_message_topic;
+
+    function setUp() public override {
+        // there seems no auto way to derive the abi in string, check this before running tests, make sure
+        // it's updated with the implementation
+        new_topdown_message_topic = keccak256(
+            "NewTopDownMessage(address,(uint8,((uint64,address[]),(uint8,bytes)),((uint64,address[]),(uint8,bytes)),uint64,uint256,bytes),bytes32)"
+        );
+
+        // get some free money.
+        vm.deal(address(this), 10 ether);
+    }
+
+    // ======= implements ipc handler =======
+
+    /* solhint-disable-next-line unused-vars */
+    function handleIpcMessage(IpcEnvelope calldata envelope) external payable returns (bytes memory ret) {
+        ret = bytes("");
+    }
+
+    receive() external payable {}
+
+    // ======= internal util methods ========
+
+    function executeTopdownMessages(IpcEnvelope[] memory msgs, GatewayDiamond gw) internal {
+        uint256 minted_tokens;
+
+        for (uint256 i; i < msgs.length; ) {
+            minted_tokens += msgs[i].value;
+            unchecked {
+                ++i;
+            }
+        }
+        console.log("minted tokens in executed top-downs: %d", minted_tokens);
+
+        // The implementation of the function in fendermint is in
+        // https://github.com/consensus-shipyard/ipc/blob/main/fendermint/vm/interpreter/contracts/fvm/topdown.rs#L43
+
+        // This emulates minting tokens.
+        vm.deal(address(gw), minted_tokens);
+
+        XnetMessagingFacet xnetMessenger = gw.xnetMessenger();
+
+        vm.prank(FilAddress.SYSTEM_ACTOR);
+        xnetMessenger.applyCrossMessages(msgs);
+    }
+
+    function createBottomUpCheckpoint(
+        SubnetID memory subnet,
+        GatewayDiamond gw
+    ) internal returns (BottomUpCheckpoint memory checkpoint) {
+        uint256 e = getNextEpoch(block.number, DEFAULT_CHECKPOINT_PERIOD);
+
+        GatewayGetterFacet getter = gw.getter();
+        CheckpointingFacet checkpointer = gw.checkpointer();
+
+        BottomUpMsgBatch memory batch = getter.bottomUpMsgBatch(e);
+
+        (, address[] memory addrs, uint256[] memory weights) = TestUtils.getFourValidators(vm);
+
+        (bytes32 membershipRoot, ) = MerkleTreeHelper.createMerkleProofsForValidators(addrs, weights);
+
+        checkpoint = BottomUpCheckpoint({
+            subnetID: subnet,
+            blockHeight: batch.blockHeight,
+            blockHash: keccak256("block1"),
+            nextConfigurationNumber: 0,
+            msgs: batch.msgs,
+            activity: ActivityHelper.newCompressedActivityRollup(1, 3, bytes32(uint256(0)))
+        });
+
+        vm.prank(FilAddress.SYSTEM_ACTOR);
+        checkpointer.createBottomUpCheckpoint(
+            checkpoint,
+            membershipRoot,
+            weights[0] + weights[1] + weights[2],
+            ActivityHelper.dummyActivityRollup()
+        );
+
+        return checkpoint;
+    }
```

Comment:
Do we really not have utils for this kind of common stuff in a test library?

Reply:
not really, I created an issue to track this good to have: https://github.com/consensus-shipyard/ipc/issues/1244